File: 362-design-hit-counter/design-hit-counter.js

var HitCounter = function() {
    this.cache = []
};

/** 
 * @param {number} timestamp
 * @return {void}
 */
HitCounter.prototype.hit = function(timestamp) {
    this.cache.push(timestamp)
};

/** 
 * @param {number} timestamp
 * @return {number}
 */
HitCounter.prototype.getHits = function(timestamp) {
    // binary search weight right
    const N = this.cache.length
    const toFind = timestamp - 300
    let l = 0
    let r = N - 1
    while (l <= r) {
        let m = Math.floor((l + r) / 2)
        if (this.cache[m] <= toFind) {
            l = m + 1
        } else {
            r = m - 1
        }
    }
    return N - l
};

/** 
 * Your HitCounter object will be instantiated and called as such:
 * var obj = new HitCounter()
 * obj.hit(timestamp)
 * var param_2 = obj.getHits(timestamp)
 */

File: 362-design-hit-counter/design-hit-counter.py
class HitCounter:

    def __init__(self):
        self.cache = []

    def hit(self, timestamp: int) -> None:
        self.cache.append(timestamp)

    def getHits(self, timestamp: int) -> int:
        # bisect right (bs weight right)
        N = len(self.cache)
        to_find = timestamp - 300
        l, r = 0, N - 1
        while l <= r:
            m = (l + r) // 2
            if self.cache[m] == to_find:
                l = m + 1
            elif self.cache[m] < to_find:
                l = m + 1
            else:
                r = m - 1
        return N - l



# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)